resteasy-core/src/main/java/org/jboss/resteasy/specimpl/BuiltResponse.java

```diff
@@ -247,9 +265,245 @@ public void setGenericType(Type genericType)
 
    public <T> T readEntity(Class<T> type, Type genericType, Annotation[] anns)
    {

```

Comment:
@rsearls, can't this code be removed from ClientResponse now?

Reply:
No it must remain.  It is used by BuiltResponseEntityNotBacked.  This class was provided to
support other select cases where entity should not be read.

Reply:
Hi @rsearls,

I'm confused. ClientResponse is derived from BuiltResponse, and it looks like this code (readEntity(), etc.) was moved from ClientResponse up to BuiltResponse. And BuiltResponseEntityNotBacked is derived from BuiltResponse ...

-Ron

Reply:
Look at prior versions of the code.  You will see that BuiltResponse line 248
has always provided the root definition for this method.

   public <T> T readEntity(Class<T> type, Type genericType, Annotation[] anns)
   {
      throw new IllegalStateException(Messages.MESSAGES.entityNotBackedByInputStream());
   }

ClientResponse overrides this method with actual code (lines 237-301)

    public synchronized <T> T readEntity(Class<T> type, Type genericType, Annotation[] anns)
       {
          abortIfClosed();
          if (entity != null)
          {
             if (type.isInstance((this.entity)))
          ................

These classes create new instances of BuiltResponse, 

    line 182  AsyncResponseConsumer
    line 471  ResourceMethodInvoker
    line 184 ResteasyView 
    line 45  MessageSanitizerMediaTypeTest

so the current code should be maintained.

Reply:
Hi @rsearls,

I'm still confused. I understand that BuiltResponseEntityNotBacked is replacing BuiltResponse because it restores the original version of readEntity(). But I see almost identical versions of readEntity() in BuiltResponse and ClientResponse, so why can't ClientResponse just inherit readEntity() from BuiltResponse? 

-Ron

Reply:
Here is more information on this.

BuiltResponse implements the 4 methods declared in Response interface.
   
   public <T> T readEntity(Class<T> type)
   public <T> T readEntity(GenericType<T> entityType)
   public <T> T readEntity(GenericType<T> entityType)
   public <T> T readEntity(GenericType<T> entityType, Annotation[] annotations)

Each of the methods above calls method 

    public <T> T readEntity(GenericType<T> entityType, Annotation[] annotations)

which is declared in BuiltResponse.

org.jboss.resteasy.client.jaxrs.internal.ClientResponse extends BuiltResponse.
This class does not override the first 4 methods.  It does override method

  public <T> T readEntity(GenericType<T> entityType, Annotation[] annotations).

org.jboss.resteasy.client.jaxrs.internal.AbortedResponse extends ClientResponse
and overrides method 

  public <T> T readEntity(GenericType<T> entityType, Annotation[] annotations)

unit-test org.jboss.resteasy.test.client.resource.NoContentStreamingCloseTestResponse 
extends BuiltResponse and overrides method 

  public <T> T readEntity(GenericType<T> entityType, Annotation[] annotations)

org.jboss.resteasy.core.ServerResponse extends BuiltResponse.  It does not
override method readEntity.  In the future it may want to override this method.

It's possible a customer may want to override this method in the future as well.

If this method is removed the first 4 methods will need to be implemented in
some way and the classes that extend from BuiltResponse might need to provide
their own implementations of these 4 methods.

Reply:
Hi @rsearls,

re: "If this method is removed the first 4 methods will need to be implemented in
some way and the classes that extend from BuiltResponse might need to provide
their own implementations of these 4 methods."

Maybe this is the problem. I meant that the implementation of that last readEntity() could be removed from ClientResponse. Maybe you think I meant it could be removed from BuiltResponse? If so,then I apologize for being unclear.

-Ron

Reply:
Currnetly this method can not simply be removed.  The key issue is that there are some
class variables that are being shared between ClientResponse and BuiltResponse and
some are not, in particular entity and bufferedEntity are causing an issue.

Do you want me to attempt to refactor this code so method readEntity can be removed from ClientResponse?